### PR TITLE
refactor: Apply new storage

### PR DESCRIPTION
**Pull request by Reto911**
Since this PR, we will handle all storage requests with the new storage.

### CODE REVIEW

File: src/new_storage/error.rs [hunk 1]
```diff
@@ -29,3 +33,35 @@ pub enum StorageError {
     #[error("{0}")]
     Internal(#[from] anyhow::Error),
 }
+
+impl From<StorageError> for DatenLordError {
+    fn from(value: StorageError) -> Self {
+        match value {
+            StorageError::OutOfRange { maximum, found } => DatenLordError::InternalErr {
```
Comment (frostyplanet) [inline src/new_storage/error.rs:40]: I think most places use StorageError::OutOfRange can be replace with assert, thus removing one error type

File: src/new_storage/backend/backend_impl.rs [hunk 2]
```diff
@@ -1,13 +1,73 @@
...
+                let mut builder = S3::default();
+
+                builder
+                    .endpoint(endpoint_url)
+                    .access_key_id(access_key_id)
+                    .secret_access_key(secret_access_key)
+                    .region("auto")
+                    .bucket(bucket_name);
+
+                Operator::new(builder)?.layer(layer).finish()
```
Comment (IRONICBo) [inline src/new_storage/backend/backend_impl.rs:58]: We may need retry and concurrency limit layer. https://github.com/datenlord/datenlord/pull/513

**Review by IRONICBo — APPROVED** [review]
LGTM

File: src/new_storage/backend/backend_impl.rs (hunk 2 above)
Comment (Reto911) [reply src/new_storage/backend/backend_impl.rs:58]: Such mechanism requires configurable arguments such as `max_concurrent_requests`, introduced in #513. Therefore, may I request you to solve this problem, when this PR merged?

Comment (IRONICBo) [reply src/new_storage/backend/backend_impl.rs:58]: Ok, I will assign myself.

**Comment by Reto911** [comment]
The new storage module has better performance on write while has worse performance on read, compared with the old one. The result of a performance test are below.

#### Write
block size (MiB) | new (MiB/s) | old (MiB/s)
-- | -- | --
0.5 | 36.4 | 30.8
1 | 33.4 | 28.1
2 | 33 | 20.2
4 | 28.8 | 14.6

#### Read
block size (MiB) | new (MiB/s) | old (MiB/s)
-- | -- | --
0.5 | 205 | 502
1 | 238 | 517
2 | 260 | 530
4 | 289 | 466

#### Read (4 jobs)
block size (MiB) | new (MiB/s) | old (MiB/s)
-- | -- | --
0.5 | 814 | 2435
1 | 871 | 2207
2 | 901 | 2634
4 | 906 | 2688

#### Read (8 jobs)
block size (MiB) | new (MiB/s) | old (MiB/s)
-- | -- | --
0.5 | 1104 | 5038
1 | 1163 | 5107
2 | 1235 | 5224
4 | 1217 | 5261

**Pull request merged**